### DISCUSSION TITLE
Help need about strange issue of x-for.

### DISCUSSION THREAD

**Discussion by leopku** [alpinejs/alpine#935] [4. General]
I'm new to Alpinejs.

There are two codes which the first works but the second sucks.

* the one worked correctly
```html
  <template x-if="$store.pokenmon" x-for="(dir, index) in $store.pokenmon.dirs" :key="index">
      <tr>
        <td x-text="dir.path"></td>
      </tr>
  </template>
```

* the one returned many blank sections
```html
    <template x-if="$store.pokenmon" x-for="(dir, index) in $store.pokenmon.dirs" :key="index">
      <section>
        <tr>
          <td x-text="dir.path"></td>
        </tr>
      </section>
    </template>
```

### Environment
* alpine@v2.8.0
* Chrome 86

How to make the second one worked as expected?

**Comment by KevinBatdorf** [comment]
That's not valid HTML so the browser itself will remove it as soon as Alpine inserts it into the dom.

**Reply by KevinBatdorf** [reply]
For example, Try adding this outside of Alpine:

```html
<section>
  <tr>
    <td>You will see this text, but no td, or tr</td>
  </tr>
</section>
```

**Reply by leopku** [reply]
Thanks for your reply.

Actually, theses codes are part of my table.

```html
        <section>
          <table>
                <thead>
                    <tr>
                        <th>head1</th>
                        <th>head2</th>
                    </tr>
                </thead>
                <section>
                  <tr>
                    <td>You will see this text, but no td, or tr</td>
                    <td>You will see this text, but no td, or tr</td>
                  </tr>
                </section>
          </table>
        </section>
```

Above code rendered a table as expected.
See a demo [here](https://codepen.io/leopku/pen/JjRNwMV)

**Reply by KevinBatdorf** [reply]
The browser is guessing the markup there as well. You can't put a `<section>` element there in a table. You can use `<tbody>` instead of section though and it will render properly. But seems you probably want just a `<tr>`?

**Reply by KevinBatdorf** [reply]
In your demo, this is what the browser is rendering it as:

<img width="329" alt="Screen Shot 2020-12-17 at 8 58 37 PM" src="https://user-images.githubusercontent.com/1478421/102496984-add57e80-40aa-11eb-9561-da09ae25304f.png">

**Reply by SimoTod** [reply]
It does not, the browser fixes it and moves tr and td outside section. When alpine kicks in, it finds 2 elements as root inside x-for which is not allowed. If you open the dev console you should be able to see a warning. Your section tag should be removed or moved inside td

**Reply by leopku** [reply]
Got it.

Thanks @KevinBatdorf and @SimoTod.

**Comment by leopku** [comment]
I changed `table` to other tags.

I really troubled with nest `x-for` now.

**Reply by KevinBatdorf** [reply]
Make sure you have a spacer:
```html
<template x-for>
  <div> <!-- spacer -->
    <template x-for>
```

**Reply by leopku** [reply]
Thank you @KevinBatdorf .


I had a nested json like

```json
{
  "path": "."
  "dirs": [
    {
      "path": "./dir001",
      "dirs": [ ... ],
      "files": [ ... ],
    },
    {
      "path": "./dir002",
      "dirs": [ ... ],
      "files": [ ... ],
    }
  ]
  "files": [
    {"name":"file001", "size": 111},
    {"name": "file002", "size": 222}
  ]
}
```

I just want to display top 2 level dirs with `details` and `summary`

* This code works
```html
    <template x-if="$store.pokenmon" x-for="(dir, index) in $store.pokenmon.dirs" :key="index">
        <details>
          <summary x-text="dir.path"></summary>
        </details>
    </template>
```

* This code broken with `dir undefined`
```html
    <template x-if="$store.pokenmon" x-for="(dir, index) in $store.pokenmon.dirs" :key="index">
        <details>
          <summary x-text="dir.path"></summary>
          <template x-for="(subdir, subindex) in dir" :key="subindex">
            <details>
              <summary x-text="subdir.path"></summary>
            </details>           
          </template>
        </details>
    </template>
```


Any ideas?

**Reply by KevinBatdorf** [reply]
Not sure, but you can only loop over an array. Maybe that's the issue here? 

https://codepen.io/KevinBatdorf/pen/rNMwjVx

**Reply by leopku** [reply]
Finally I found the issue by compare words by words between your codes and mine.
The only difference between us is just `section` and `div`. 🤣 

After changing `section` to `div`, I got what I want.

I don't know `x-data` in `section` tag not working sometimes is a feature or a bug?

**Reply by SimoTod** [reply]
It sounds weird. Section is pretty much a div with a semantic meaning.
It works correctly to me: https://codepen.io/SimoTod/pen/VwKWpbO
